### ISSUE TITLE
Mouse Wheel Zoom In/Out

### ISSUE THREAD

**Issue by nvdl** [lanceman2/quickplot#4]
Good to have it.

**Comment by lanceman2** [comment]
I take it this is a feature request.

So like in google maps?  I just looked at google maps.  I like it.  I seems like it would be easy to do, at least for me.  I can't say when I'll get it done.  2 weeks maybe.  I just need a free hour or 2.

So the zoom is keeping the aspect (x/y) and keeping the point of the pointer with the same point of the image as it zooms.   Zooms in (magnify) for wheel up and zooms out for wheel down.    Is that correct?

--lance

**Comment by lanceman2** [comment]
My reply.
https://github.com/lanceman2/quickplot/issues/4

I like it.
--lance

**Comment by nvdl** [comment]
I am thinking if mouse and key bindings can be in a settings file so one can edit them.

> Zooms in (magnify) for wheel up and zooms out for wheel down. Is that correct?

Yes, that is correct.

Moreover, I was unable to find cursors to measure a pulse length from a scope trace.

**Comment by lanceman2** [comment]
Setting file is a different issue, I made another ticket for that.  #5

I don't follow this: "I was unable to find cursors to measure a pulse length from a scope trace."
--lance

**Comment by lanceman2** [comment]
I wrote an oscilloscope program too, at  https://github.com/lanceman2/quickscope
No one knows about it.  It's way way better than Xscope.

**Comment by nvdl** [comment]
> I don't follow this: "I was unable to find cursors to measure a pulse length from a scope trace."

If get a CSV from a scope, I want to measure the time range for a certain event just like a cursor on the scope screen.

Quickscope look very nice.
Gave it a short look.
Does it support loading trace data from a file?

If Quickscope can do all of the plotting, why Quickplot?
Am I missing something?

**Comment by lanceman2** [comment]
> Does it support loading trace data from a file?

 Yes, but it requires adding support for your file format, whatever it is.  Quickscope is an API (application programming interface) library that comes with many programs.

>If Quickscope can do all of the plotting, why Quickplot?
Am I missing something?

Yes you are missing something:  Quickscope displays data as it comes in.  It tries to view the data as soon as possible.  It's like a real Oscilloscope.  The lines are animated (in time), without user interaction (where as quickplot is static).  The delay from input to viewing the data is small.  The delay time from input to viewing is in the order of monitor draw time, like 1/60 second.  Quickscope only buffers the data that is being displayed, where as quickplot buffers the whole file and the data is not changing with time after it is loaded.

Both quickplot and quickscope can display a file of time series data, but quickplot just loads the whole file and displays it (it's static), and quickscope displays the file like it's playing it back.  Quicksocpe can read your microphone, and keep reading and displaying it forever.  Quickplot can only read finite microphone input captures.

The most common example file is a sound file.  Quickplot gives a static view of the whole file.  Quickscope gives a visual play-back of the file which you can sync with a speaker.  I do that all the time.  Listen (sound player) and watch (quickscope) a sound file.

Quickplot and quickscope can be used in a complementary fashion and I'd like to make better interfaces between them.

I know I'm just talking to myself at this point, but I'm generating ideas ...  Thanks for the spark.

**Comment by nvdl** [comment]
> Quickscope only buffers the data that is being displayed, where as quickplot buffers the whole file and the data is not changing with time after it is loaded.

I think Quickscope can do all the things if it supports loading part of a data file in which Quickplot may not be needed. It's like freezing the scope from triggering and the screen showing the last drawn buffer statically.

So, if there is a file buffer module, it can feed Quickplot.
For loading data from huge files where a small area is zoomed in, loading partial data will speed up plotting. I had done a plotter like this a while ago where one could choose the x-range and the plotter would seek and draw from that location in the file.

**Comment by lanceman2** [comment]
I like the way you think, and I've tried to make it all one program, but the internal data structures in the two programs end up very different.  In the user space memory: Quickscope has a special x-y pixmap that is fed directly from the input, the data is converted to the pixmap in small chunks at a time and it has no record as to what the original input data was, were as quickplot keeps all the input data as a list of doubles and renders it to a x-y pixmap at the time it draws at a given zoom level.  Quickplot loads all the data before drawing (every time), and quickscope does not, it draws data as it is read.  It's (quickplot) 1 chunk vs. (quickscope) infinite number of chunks.  Quickplot, at best, can be a scope capture viewer, if I add another buffer to capture the input as a list of doubles (floats of some kind) to quickscope.   That (capture mode) would kill quickscope performance, so it would have to be optional.   
Quickscope does have a built in screen capture, but only as an image. 

Quickscope is trying to be fast; it does a minimum of buffering.  Think of it having each pixel mapped to a given input value with no record of what the value was (it only keeps color and a pointer to prev and next pixel), and that pixel is fading in time (which is why it needs prev and next), and it has any number of channels feeding each pixel, at any time.  It has any number ot adjustable filters for each channel.  It's a generic temporally ordered pixel color manager.   Quickplot stores channels of data.  Quickscope does not, and it can't, at least in it's core (fast running) design.

Note: triggering is not required for an oscilloscope.  Oscilloscope's have free-run and X-Y modes that usually are not triggered.  Quickscope is more like a real scope because of that.  Or an old CRT scope anyway.

I'm using a very different approach than most other scope programs.  It should be a fast scope, given that the drawing API that I build it on is fast.

>loading partial data will speed up

I can see that as being very useful for large files, assuming that the data is ordered a certain way.  I have thought a lot about different culling strategies.   Currently I never have files bigger than a few Megabytes (for quickplot).  I can imagine that a sound track (and/or video track) editor could use a pool of memory mapped file chunks, which may be a common design pattern.  I've been there on other projects.   Currently quickplot only culls for function plots by keeping two end pointers to the X field, at a given zoom level (an in-memory culling method).  That can't work for non-function plots.  Quickplot works with more general unordered data (scatter plot for example) too, and so does quickscope.

**Bottom line:**
It is possible to build all the functionality of quickplot with quickscope, and the result needs to be called "SlowPlot".    The vice versa is called SlowScope.  They are currently two extreme cases in buffering and real-time-ish-ness (word).


Please let me know if you actually run quickscope.  I'm the only one who has thus far (that I know).

--lance

**Comment by nvdl** [comment]
I agree with the idea behind Quickscope being as fast as possible but that is good for waveforms that are not to be looked into detail. Beyond 60fps (or possibly less), I feel there is no gain.
If Quickscope can draw without dropping any sample from a sampler in real time, it will be helpful e.g. for MCU based ADC scopes connected via USB.

I have one basic scope written in Pygame (to use SDL) running at slow sample rate. High sample rate is not needed (and undesirable) in that case.

>  Oscilloscope's have free-run and X-Y modes that usually are not triggered

By triggering, I meant modern scopes where the trace is preserved on screen.
So it is more like conditional halt based on waveform characteristics.

Have not looked into Quickscope yet. I was thinking for a unified solution where one can view and importantly analyze a waveform. The waveform data can come from a saved file or a data buffer when a trigger is detected. The scope can be free running but I don't think free running data analysis is needed in that case.
